MCR-2604 OCFL Classification Storage

Hammer1279 opened the pull request:
This adds a Classification Storage in a OCFL Repository, while the Database handles access to the files, the OCFL Store works like a backup and stores changes versioned. 

In additon, to work, this also brings an DAO Implementation with added Events and a ClassificationManager Interface.

[Link to jira](https://mycore.atlassian.net/browse/MCR-2604).

Hammer1279 committed:
MCR-2604 OCFL Classification Storage

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

Hammer1279 commented:
Docs can be found here: https://github.com/MyCoRe-Org/mycore-website/blob/ocfl-class/mycore.org/content/de/documentation/basics/basics_mcr_ocfl.md

yagee-de commented on mycore-ocfl/src/main/java/org/mycore/ocfl/MCROCFLPersistenceTransaction.java:
```diff
@@ -0,0 +1,165 @@
+     */
+    @Override
+    public boolean getRollbackOnly() {
+        if (verboseLogging) {
```
LOGGER.debug already checks if Specific logger is on debug level. No need to specify any property or adding custom checks

yagee-de commented on mycore-base/src/main/java/org/mycore/datamodel/classifications2/impl/MCREventedCategoryDAOImpl.java:
```diff
@@ -0,0 +1,127 @@
+    protected void queueForCommit(MCREvent evt) {
+        String classQueue = "classQueue";
+        MCRSession currentSession = MCRSessionMgr.getCurrentSession();
+        ((ArrayList<MCREvent>) currentSession.get(classQueue)).add(evt);
```
String "classQueue" is used to often in Code. This should be centralized somewhere.
Also a BIG no go is to use session as a storage for transactional data without any transactional reference. There could be two different transactions running concurrently and one is rolled back. In this case all the data of two transactions are lost.

yagee-de commented on mycore-ocfl/src/main/java/org/mycore/ocfl/MCROCFLXMLClassificationManager.java:
```diff
@@ -0,0 +1,332 @@
+        MCRCategoryID mcrid = (MCRCategoryID) data.get("mid");
+        MCRCategory mcrCg = (MCRCategory) data.get("ctg");
+        MCRContent rtXml = (MCRContent) data.get("rtx"); // class
+        MCREvent event = (MCREvent) evt.get("event");
```
Do not use magic Strings here. Make them "public static final" somewhere and document them. What is that "event" within an event anyway?

yagee-de commented on mycore-ocfl/src/main/resources/components/ocfl/config/mycore.properties:
```diff
@@ -16,14 +16,73 @@
+#   / /  / // /_/ // /___ / /_/ // _, _//  __/  / /_/ // /___ / __/  / /___
+#  /_/  /_/ \__, / \____/ \____//_/ |_| \___/   \____/ \____//_/    /_____/
+#          /____/                                                          
+
```
please: no!

Hammer1279 committed:
MCR-2604-ocfl-classification-storage
Remove text and useless code

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

yagee-de commented:
Properties of a transaction are:

- Atomicity
- Consistency
- Isolation
- Durability

Atomicity is hard to achieve on a filesystem storage, we could rollback to a previous version. So not that big of a deal.
Consistency seems not to be harmed by this PR.
Isolation is currently weakened as any change is visible to any other user in a mutable directory and transaction commits data from other transactions within the same session. Also different transaction could write to the same resource.
Durability is due to ocfl.

yagee-de commented:
You could keep your data in a transaction object and have that transaction being committed later. If you NEED to use a mutable HEAD of an OCFL object, make sure to acquire a lock to operate on that. That lock should belong to the transaction and if the transaction is committed than "commitStagedChanges" on that object. If a rollback occurs, free the lock after you "purgeStagedChanges".

yagee-de commented:
Thanks for your documentation and your work. I really appreciate your push towards a OCFL storage option.

Hammer1279 committed:
MCR-2604-ocfl-classification-storage
Update all kinds of code

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

yagee-de committed:
MCR-2604 work in progress with ThreadLocal

Hammer1279 committed:
MCR-2604-ocfl-classification-storage
Use ThreadLocal instead of MutableHead

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

Hammer1279 committed:
MCR-2604-ocfl-classification-storage
Update Properties

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

Hammer1279 committed:
MCR-2604-ocfl-classification-storage
remove @Nullable annotations to avoid adding a unnecessary dependency

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

Hammer1279 committed:
MCR-2604-ocfl-classification-storage
Code Cleanup

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

yagee-de committed:
MCR-2604 remove usage of Optional in field and fixed warnings

yagee-de committed:
MCR-2604 add new line at end of file

yagee-de committed:
MCR-2604 fixed method naming and removed redundant code

yagee-de commented on mycore-ocfl/src/main/java/org/mycore/ocfl/commands/MCROCFLDevCommands.java:
```diff
@@ -0,0 +1,124 @@
+        } catch (Exception e) {
+            // throw new IOException(e);
+            // expected Error incase the directory doesn't exist yet
+        }
```
Does not use OCFL-API and requires a specific non-default storage layout

yagee-de commented on mycore-ocfl/src/main/java/org/mycore/ocfl/MCROCFLXMLClassificationManager.java:
```diff
@@ -0,0 +1,213 @@
+        } else {
+            return rootFolder + mcrid.getRootID() + '/' + mcrid.toString() + ".xml";
+        }
+    }
```
Where are all these files come from? Where is this specified? How to we ensure that if a classification is deleted all categories are deleted, too?

yagee-de commented on mycore-ocfl/src/main/java/org/mycore/ocfl/MCROCFLXMLClassificationManager.java:
```diff
@@ -0,0 +1,213 @@
+
+    }
+
+    public void fileDelete(MCRCategoryID mcrid) {
```
method naming: We have "create", "update" and "fileDelete". Should be renamed to "delete"

yagee-de commented:
We should have a discussion about messages beeing parsed to determine what action caused a revision. This does not look good to me.

Hammer1279 commented on mycore-ocfl/src/main/java/org/mycore/ocfl/MCROCFLPersistenceTransaction.java:
```diff
@@ -87,12 +82,12 @@ public void commit() {
         }
         final Map<MCRCategoryID, MCRCategory> mapOfChanges = CATEGORY_WORKSPACE.get();
-        final MCROCFLXMLClassificationManager ocflClassficationManager = managerOpt.get();
+        final MCROCFLXMLClassificationManager ocflClassficationManager = manager;
```
Why not just change all occurances of `ocflClassificationManager` to `manager`?

Hammer1279 committed:
MCR-2604-ocfl-classification-storage
Update Commands, PersistanceTransaction and ClassificationManager

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

yagee-de committed:
MCR-2604 always use classification subdirectory

remove unused properties

yagee-de committed:
MCR-2604 split commands to reuse code better

yagee-de committed:
MCR-2604 minor refactoring

yagee-de committed:
MCR-2604 minor refactoring

yagee-de committed:
MCR-2604 rename "repair" command to "sync"

Hammer1279 committed:
MCR-2604-ocfl-classification-storage
update thrown error type and message

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

yagee-de committed:
MCR-2604 introduce MCROCFLObjectIDPrefixes

yagee-de committed:
MCR-2604 use event types as characters

read classification as late as possible ans remove redundant reads
minor refactorings

yagee-de committed:
MCR-2604 renamed objName to ocflObjectID

yagee-de committed:
MCR-2604 use IOException more often

yagee-de committed:
MCR-2604 removed unused fields

yagee-de committed:
MCR-2604 renamed event handler class

Hammer1279 committed:
MCR-2604 fixed MCRStorageLayout

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

yagee-de committed:
MCR-2604 fixed NPE

yagee-de committed:
MCR-2604 fix modified date
